src/java/org/apache/cassandra/index/sai/disk/PrimaryKeyWithSource.java

```diff
@@ -110,21 +138,31 @@ public boolean equals(Object o)
         if (o instanceof PrimaryKeyWithSource)
         {
             var other = (PrimaryKeyWithSource) o;
+            // If they are from the same source sstable, we can compare the row ids directly.
             if (sourceSstableId.equals(other.sourceSstableId))
                 return sourceRowId == other.sourceRowId;
         }
-        return primaryKey.equals(o);
+
+        if (!(o instanceof PrimaryKey))
+            return false;
+
+        var other = (PrimaryKey) o;
+        // If the source sstable does not contain the PrimaryKey, the keys cannot be equal.
+        if (sourceSstableMinKey.compareTo(other) > 0 || sourceSstableMaxKey.compareTo(other) < 0)
+            return false;
+
+        return primaryKey().equals(other);
     }
 
     @Override
     public int hashCode()
     {
-        return primaryKey.hashCode();
+        return primaryKey().hashCode();
     }
 
     @Override
     public String toString()
     {
-        return String.format("%s (source sstable: %s, %s)", primaryKey, sourceSstableId, sourceRowId);
+        return String.format("%s (source sstable: %s, %s)", primaryKey(), sourceSstableId, sourceRowId);
```

Comment:
this is implicitly loading the primary key, I hope that we don't fall into this trap in some logging statement, is it worth to not force loading the primary if it is not loaded and report only the sources ?

Reply:
Sure, since we have the sstable and row id, we probably don't need to force the load.